runtime/eventlog.c

```diff
@@ -0,0 +1,257 @@
+#define CAML_INTERNALS
+#include <stdio.h>
+#include <string.h>
+#include <stdint.h>
+#include <unistd.h>
+#include <inttypes.h>
+#include "caml/alloc.h"
+#include "caml/eventlog.h"
+#include "caml/osdeps.h"
+
+#define CTF_MAGIC 0xc1fc1fc1
+#define CAML_TRACE_VERSION 0x1
+
+struct ctf_stream_header {
+  uint32_t magic;
+  uint16_t caml_trace_version;
+  uint16_t stream_id;
+};
+
+static struct ctf_stream_header header = {
+  CTF_MAGIC,
+  CAML_TRACE_VERSION,
+  0
+};
+
+#pragma pack(1)
+struct ctf_event_header {
+  uint64_t timestamp;
+  uint32_t pid;
+  uint32_t id;
+};
+
+static uintnat alloc_buckets [20] =
+  {0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0};
+
+
+struct event {
+  struct ctf_event_header header;
+  uint8_t  phase; /* for GC events */
+  uint8_t  counter_kind; /* misc counter name */
+  uint8_t  alloc_bucket; /* for alloc counters */
+  uint32_t count; /* for misc counters */
```

Comment:
I wonder if the size restrictions here are future-proof. Wouldn't it be safer to use int16 for `phase`, `counter_kind` and `alloc_bucket`? For "misc counters", can't we have above-32bit values so wouldn't int64 be safer? (Eg. memory usage counters for large programs.) You already use 128bits of header metadata, so it's not clear that int8 vs. int16 is so important there.

Reply:
(Note for readers: `phase` and `counter_kind` here refer not just to toplevel notions of phase/kind (mark, sweep, clean, etc.), but also the name finer subdivisions; there is in fact a hierarchy of moments with a tree structure. It seems fairly conceivable that we would end up with more than 2^8 "counter kinds" if we extend to support more runtime tracing.

Reply:
I think counters should be pushed to 64bits indeed but I think the packet header could be shortened further and made more efficient. (so I think the payload size topic is not quite irrelevant) 
For instance the `pid` field can be removed and put into the stream header itself, since for now we do not plan to have multiple process/domains emitting to a same stream. (rather, I believe it make more sense to have a separate stream per process/domain).

As for the phase and counter kind, I think you are right.
I'm wondering if this is the right categorization model however, surely an ever-growing enumeration is not adviceable? Maybe things could be sorted in a better way and split in various enumerations.
I added to the repository a few days ago the TSDL declaration, to make discussions on the schema a bit simpler. :)

Reply:
From a high-level perspective a natural categorization I think would be hierarchical, with "intervals" as nodes in the tree and "events" as child-less leaves. But It's not clear to me whether the event format supports this, and given that the killer feature is the availability of nice visualization tools we don't want to go for something non-standard. Flattening all interval and event names into basically a single enum, as you did for GC instrumentation, may be a way to go to fit the common format -- and then we can indeed expect an explosion of enumeration size.